Allow private keys to come from an encrypted variable for secure portability

scottbrown opened the pull request:
Hi,

The issue with specifying the private key as a file is that the file could mistakenly be kept inside the version control system and Ansible doesn't provide the ability to read encrypted files so the file must be kept unencrypted.

AFAIK, the Ansible best practice is to store the sensitive data as a variable (if possible) and encrypt the variable file.  If the file is kept outside of the project, then the user of the project must find the file and load it into their system.  This is again insecure and prone to failure.  This PR offers Ansible users a more secure option with almost no learning curve or project tribal knowledge.

Here is a playbook where the key contents are specified as an encrypted variable, which is loaded at runtime with a variable file encrypted with Ansible Vault.

``` Yaml

---
- hosts: all
  gather_facts: no
  connection: local
  tasks:
    - include_vars: privatekey_encrypted.yml
  - name: add certificate using encrypted variable
    iam_cert:
      name: 'star.example.com'
      cert: "{{ playbook_dir }}/files/star.example.com.crt"
      key_contents: "{{ privatekey_encrypted }}"
      state: present
```

The variable file `privatekey_encrypted.yml` looks like this if it were viewed unencrypted:

``` Yaml

---
privatekey_encrypted: |
  -----BEGIN RSA PRIVATE KEY-----
  ...
  -----END RSA PRIVATE KEY-----
```

Since `key` and `key_contents` are mutually exclusive, there are validation check built into the module that will cause a graceful failure.

``` Yaml

---
- hosts: all
  gather_facts: no
  connection: local
  tasks:
    - include_vars: privatekey_encrypted.yml
    - name: add certificate using file and variable
      iam_cert:
        name: 'star.example.com'
        cert: "{{ playbook_dir }}/files/star.example.com.crt"
        key: "{{ playbook_dir }}/files/star.example.com.key"
        key_contents: "{{ privatekey_plaintext }}"
        state: present
```

This results in the following failure during runtime:

```
$ ansible-playbook -i localhost.inventory add_cert.yml

PLAY ***************************************************************************

TASK [include_vars] ************************************************************
ok: [localhost]

TASK [add certificate using encrypted variable] ********************************
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "key or key_con
tents can be specified, not both."}

PLAY RECAP *********************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1
```

scottbrown committed:
Adding the ability to specify the private key as coming from an encrypted variable for secure portability

scottbrown committed:
Fixing missing colon.  Hit the backspace before saving.

gregdek commented:
Thanks @scottbrown. @defionscode please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit' or 'needs_revision' as appropriate.

robynbergeron commented:
@defionscode This change is still pending your review; do you have time to take a look and comment? Please comment with text 'shipit' or 'needs_revision' as appropriate.

gregdek commented:
@defionscode still waiting on your review.  Please comment with text 'shipit' or 'needs_revision' as appropriate. If we don't hear from you within 14 days, we will start to look for additional maintainers for this module.

gregdek commented:
Thanks @scottbrown for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

juliedavila commented on cloud/amazon/iam_cert.py:
```diff
@@ -228,6 +249,7 @@ def main():
+        key_contents=dict(default=None, required=False),
         cert_chain=dict(default=None, required=False),
         new_name=dict(default=None, required=False),
         path=dict(default='/', required=False),
```
Instead of the failure occurring inside of  `def private_key_from_file_or_inline(module):`

the failure should be caught in the argument_spec as so

``` python

argment_spec.update(dict(
    state=dict...
    #rest of the spec
),
mutually_exclusive=[['key','key_contents']]
)

```

scottbrown replied:
I didn't know `mutually_exclusive` existed!  Thanks for the insight.  I'll get this patched up and ready to go.

juliedavila replied:
no worries, let us know when it's ready

[image: photo] http://linkedin.com/in/davilaio
http://twitter.com/defionscode  http://github.com/defionscode Jonathan
I. Davila
Principal Architect, Automation Practice, Red Hat, Inc
202.415.1878 // redhat.com // jdavila@redhat.com // Washington, D.C

On Fri, Mar 11, 2016 at 5:03 PM, Scott Brown notifications@github.com
wrote:

> In cloud/amazon/iam_cert.py
> https://github.com/ansible/ansible-modules-core/pull/2976#discussion_r55895517
> :
> 
> > @@ -228,6 +249,7 @@ def main():
> >          name=dict(default=None, required=False),
> >          cert=dict(default=None, required=False),
> >          key=dict(default=None, required=False),
> > -        key_contents=dict(default=None, required=False),
> >        cert_chain=dict(default=None, required=False),
> >        new_name=dict(default=None, required=False),
> >        path=dict(default='/', required=False),
> 
> I didn't know mutually_exclusive existed! Thanks for the insight. I'll
> get this patched up and ready to go.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ansible/ansible-modules-core/pull/2976/files#r55895517
> .

juliedavila commented:
I agree with the intent of this PR. Just fixup what I commented and it should be good to go

scottbrown committed:
Refactoring to use provided mutually exclusive attribute

scottbrown committed:
Fixing linting errors on new code

scottbrown commented:
All fixed.  Let me know if there's anything you need.

psykotox commented:
It would be perfect if we can use encrypted (or not) variable for cert and cert_chain parameters.

psykotox commented:
This is my crappy code to do that (line 265) : 
`
        try:
            cert = open(module.params.get('cert'), 'r').read().rstrip()
        except:
            cert = module.params.get('cert').rstrip()
        try: 
            key = open(module.params.get('key'), 'r').read().rstrip()
        except: 
            key = module.params.get('key').rstrip()
        if cert_chain is not None:
            try: 
                cert_chain = open(module.params.get('cert_chain'), 'r').read()
            except:
                cert_chain = module.params.get('cert_chain')
`

psykotox commented:
This is my crappy code to do that (line 265) : 

``` python
        try:
            cert = open(module.params.get('cert'), 'r').read().rstrip()
        except:
            cert = module.params.get('cert').rstrip()
        try: 
            key = open(module.params.get('key'), 'r').read().rstrip()
        except: 
            key = module.params.get('key').rstrip()
        if cert_chain is not None:
            try: 
                cert_chain = open(module.params.get('cert_chain'), 'r').read()
            except:
                cert_chain = module.params.get('cert_chain')
```

gregdek commented:
Thanks @scottbrown for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

scottbrown committed:
Fixing merge conflict with upstream

scottbrown commented:
I fixed the merge conflict with upstream [1] but the Travis CI is failing, but not due to the code that I changed.  Maybe something new in `devel` caused it to fail?

[1] Merge conflict caused by `devel` shifting over time.  This PR has been open for awhile.

gregdek commented:
@scottbrown A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

scottbrown committed:
Merge branch 'devel' into add-encrypted-private-key-contents

scottbrown commented:
ready_for_review

I also need some help here.  I synced my branch with upstream twice now and everything merged properly on my end (or else I wouldn't be able to push).  Yet the TravisCI still thinks there's a merge conflict.  I'm not sure what to do here.  What am I missing on my side, since my Git client doesn't show any merge conflict?  Even Github is saying that there are no conflicts with the base branch, so maybe the issue is within TravisCI itself?

gregdek commented:
Thanks @scottbrown for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]
